File: python/raspberryPi/DFRobot_GP85XX.py
# -*- coding: utf-8 -*
'''!
  @file  DFRobot_GP85XX.py
  @brief
  @copyright  Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license  The MIT License (MIT)
  @author  [Baikunlin](kunlin.bai@dfrobot.com)
  @version  V1.0
  @date  2023-02-10
  @url  https://github.com/DFRobot/DFRobot_GP85XX
'''

from __future__ import print_function
import smbus
import time
import datetime
import RPi.GPIO as GPIO

## i2c address
DF2301Q_I2C_ADDR = 0x58
##Select DAC output voltage of 0-2.5V
OUTPUT_RANGE_5V = 0
##Select DAC output voltage of 0-VCC
OUTPUT_RANGE_10V = 17
##Select DAC output voltage of 0-2.5V
OUTPUT_RANGE_2_5V = 0
##Select DAC output voltage of 0-VCC
OUTPUT_RANGE_VCC = 17


class DFRobot_GP85XX(object):
  def __init__(self):
    '''!
      @brief Module init
    '''
    pass


class DFRobot_GP8503(DFRobot_GP85XX):
  ## Configure sensor register
  GP8503_CONFIG_CURRENT_REG = 0x02
  ## Store function timing start head
  GP8503_STORE_TIMING_HEAD = 0x02
  ## The first address for entering store timing
  GP8503_STORE_TIMING_ADDR = 0x10
  ## The command 1 to enter store timing
  GP8503_STORE_TIMING_CMD1 = 0x03
  ## The command 2 to enter store timing
  GP8503_STORE_TIMING_CMD2 = 0x00
  ## Total I2C communication cycle 5us
  I2C_CYCLE_TOTAL = 0.000005
  ## The first half cycle of the total I2C communication cycle 2us
  I2C_CYCLE_BEFORE = 0.000002
  ## The second half cycle of the total I2C communication cycle 3us
  I2C_CYCLE_AFTER = 0.000003

  GP8503_STORE_TIMING_DELAY = 0.000010

  def __init__(self, addr=DF2301Q_I2C_ADDR):
    self._addr = addr
    self.outPutSetRange = 0x01
    self.voltage = 5000
    self._scl = 3
    self._sda = 2
    self.dataTransmission = 0
    GPIO.setmode(GPIO.BCM)
    GPIO.setwarnings(False)
    self.i2c = smbus.SMBus(1)

  def begin(self):
    '''!
      @param Initialize the sensor
    '''
    if self.i2c.read_byte(self._addr) != 0:
      return 0
    return 1

  def set_DAC_outrange(self, mode, vcc=5):
    '''!
      @brief 设置DAC输出范围
      @param VCC 输入的电源电压
      @param mode DAC输出范围
    '''
    if mode == OUTPUT_RANGE_2_5V:
      self.voltage = 2500
    elif mode == OUTPUT_RANGE_VCC:
      self.voltage = vcc * 1000

  def set_DAC_out_voltage(self, data, channel, value=4095):
    '''!
      @brief 设置不同通道输出DAC值
      @param data 需要输出的电压值
      @param channel 输出通道 0:通道0;1:通道1;2:全部通道
      @param value DAC分辨率 4096 : 12bit ; 32767 : 15bit
    '''

    data = int((float(data) / self.voltage) * value)
    if value == 4095:
      data = int(data) << 4
      self._send_data(data, channel)
    else:
      self._send_data(data, 0)

  def store(self):
    '''!
      @brief   将设置的电压保存在芯片内部
    '''
    GPIO.setup(self._scl, GPIO.OUT)
    GPIO.setup(self._sda, GPIO.OUT)
    self._start_signal()
    self._send_byte(self.GP8503_STORE_TIMING_HEAD, 0, 3, False)
    self._stop_signal()
    self._start_signal()
    self._send_byte(self.GP8503_STORE_TIMING_ADDR)
    self._send_byte(self.GP8503_STORE_TIMING_CMD1)
    self._stop_signal()

    self._start_signal()
    self._send_byte(self._addr << 1, 1)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2, 1)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2, 1)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2, 1)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2, 1)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2, 1)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2, 1)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2, 1)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2, 1)
    self._stop_signal()

    time.sleep(self.GP8503_STORE_TIMING_DELAY)

    self._start_signal()
    self._send_byte(self.GP8503_STORE_TIMING_HEAD, 0, 3, False)
    self._stop_signal()
    self._start_signal()
    self._send_byte(self.GP8503_STORE_TIMING_ADDR)
    self._send_byte(self.GP8503_STORE_TIMING_CMD2)
    self._stop_signal()

  def _send_data(self, data, channel):
    if channel == 0:
      self.i2c.write_word_data(self._addr, self.GP8503_CONFIG_CURRENT_REG, data)
    elif channel == 1:
      self.i2c.write_word_data(self._addr, self.GP8503_CONFIG_CURRENT_REG << 1, data)
    else:
      self.i2c.write_word_data(self._addr, self.GP8503_CONFIG_CURRENT_REG, data)
      self.i2c.write_word_data(self._addr, self.GP8503_CONFIG_CURRENT_REG << 1, data)

  def _start_signal(self):
    GPIO.output(self._scl, GPIO.HIGH)
    GPIO.output(self._sda, GPIO.HIGH)
    time.sleep(self.I2C_CYCLE_BEFORE)
    GPIO.output(self._sda, GPIO.LOW)
    time.sleep(self.I2C_CYCLE_AFTER)
    GPIO.output(self._scl, GPIO.LOW)
    time.sleep(self.I2C_CYCLE_TOTAL)

  def _stop_signal(self):
    GPIO.output(self._sda, GPIO.LOW)
    time.sleep(self.I2C_CYCLE_BEFORE)
    GPIO.output(self._scl, GPIO.HIGH)
    time.sleep(self.I2C_CYCLE_TOTAL)
    GPIO.output(self._sda, GPIO.HIGH)
    time.sleep(self.I2C_CYCLE_TOTAL)

  def _recv_ack(self, ack=0):
    ack_ = 0
    error_time = 0
    GPIO.setup(self._sda, GPIO.IN)
    time.sleep(self.I2C_CYCLE_BEFORE)
    GPIO.output(self._scl, GPIO.HIGH)
    time.sleep(self.I2C_CYCLE_AFTER)
    while GPIO.input(self._sda) != ack:
      time.sleep(0.000001)
      error_time += 1
      if error_time > 250:
        break
    ack_ = GPIO.input(self._sda)
    time.sleep(self.I2C_CYCLE_BEFORE)
    GPIO.output(self._scl, GPIO.LOW)
    time.sleep(self.I2C_CYCLE_AFTER)
    GPIO.setup(self._sda, GPIO.OUT)
    return ack_

  def _send_byte(self, data, ack=0, bits=8, flag=True):
    i = bits
    data = data & 0xFF
    while i > 0:
      i -= 1
      if data & (1 << i):
        GPIO.output(self._sda, GPIO.HIGH)
      else:
        GPIO.output(self._sda, GPIO.LOW)
      time.sleep(self.I2C_CYCLE_BEFORE)
      GPIO.output(self._scl, GPIO.HIGH)
      time.sleep(self.I2C_CYCLE_TOTAL)
      GPIO.output(self._scl, GPIO.LOW)
      time.sleep(self.I2C_CYCLE_AFTER)
    if flag:
      return self._recv_ack(ack)
    else:
      GPIO.output(self._sda, GPIO.LOW)
      GPIO.output(self._scl, GPIO.HIGH)
    return ack


class DFRobot_GP8512(DFRobot_GP85XX):

  def __init__(self):

    self.gp8503 = DFRobot_GP8503(DF2301Q_I2C_ADDR)

  def begin(self):
    '''!
     @brief 初始化函数
    '''
    return self.gp8503.begin()

  def set_DAC_outrange(self, mode, vcc=5):
    '''!
      @brief 设置DAC输出范围
      @param VCC 输入的电源电压
      @param mode DAC输出范围
    '''
    if mode == OUTPUT_RANGE_2_5V:
      self.gp8503.set_DAC_outrange(mode)
    elif mode == OUTPUT_RANGE_VCC:
      self.gp8503.set_DAC_outrange(mode, vcc)

  def set_DAC_out_voltage(self, data):
    '''!
      @brief 设置输出DAC值
      @param data 需要输出的电压值
    '''
    self.gp8503.set_DAC_out_voltage(data, 0, 32767)

  def store(self):
    '''!
      @brief   将设置的电压保存在芯片内部
    '''
    self.gp8503.store()


class DFRobot_GP8501(DFRobot_GP85XX):

  def __init__(self):
    self.voltage = 5000
    self.R0 = 18
    self.R1 = 12

  def begin(self):
    '''!
      @brief 初始化函数
    '''
    GPIO.setwarnings(False)  # 屏蔽警告
    GPIO.setmode(GPIO.BCM)  # 设置引脚模式BCM
    GPIO.setup(self.R0, GPIO.OUT)  # 设置引脚R0方向输出
    GPIO.setup(self.R1, GPIO.OUT)  # 设置引脚R1方向输出

  def set_DAC_outrange(self, mode, frequency, channel, vcc=5):
    '''!
      @brief 设置DAC输出范围
      @param mode DAC输出范围
      @param frequency PWM频率
      @param channel 输出通道 0:通道0;1:通道1;2:全部通道
      @param vcc VCC输入的电源电压
    '''
    if mode == OUTPUT_RANGE_2_5V:
      self.voltage = 2500
    elif mode == OUTPUT_RANGE_VCC:
      self.voltage = vcc * 1000
    if channel == 0:
      self.PR0 = GPIO.PWM(self.R0, frequency)
      self.PR0.start(0)
    elif channel == 1:
      self.PR1 = GPIO.PWM(self.R1, frequency)
      self.PR1.start(0)
    else:
      self.PR0 = GPIO.PWM(self.R0, frequency)
      self.PR0.start(0)
      self.PR1 = GPIO.PWM(self.R1, frequency)
      self.PR1.start(0)

  def set_DAC_out_voltage(self, data, channel):
    '''!
      @brief 设置不同通道输出DAC值
      @param data 需要输出的电压值
      @param channel 输出通道 0:通道0;1:通道1;2:全部通道
    '''
    data = int(data / self.voltage) * 100

    if data > 100:
      data = 100
    if data < 0:
      data = 0
    if channel == 0:
      self.PR0.ChangeDutyCycle(data)
    elif channel == 1:
      self.PR1.ChangeDutyCycle(data)
    else:
      self.PR0.ChangeDutyCycle(data)
      self.PR1.ChangeDutyCycle(data)


class DFRobot_GP8101(DFRobot_GP85XX):

  def __init__(self):
    self.voltage = 5000
    self.R = 18

  def begin(self):
    '''!
      @brief 初始化函数
    '''
    GPIO.setwarnings(False)  # 屏蔽警告
    GPIO.setmode(GPIO.BCM)  # 设置引脚模式BCM
    GPIO.setup(self.R, GPIO.OUT)  # 设置引脚18方向输出

  def set_DAC_outrange(self, mode, frequency):
    '''!
      @brief 设置DAC输出范围
      @param mode DAC输出范围
      @param frequency PWM频率
    '''
    if mode == OUTPUT_RANGE_5V:
      self.voltage = 5000
    elif mode == OUTPUT_RANGE_10V:
      self.voltage = 10000
    self.PR = GPIO.PWM(self.R, frequency)
    self.PR.start(0)

  def set_DAC_out_voltage(self, data):
    '''!
      @brief 设置输出DAC值
      @param data 需要输出的电压值
    '''
    data = data / self.voltage * 100
    if data > 100:
      data = 100
    if data < 0:
      data = 0
    self.PR.ChangeDutyCycle(data)

File: DFRobot_GP85XX.cpp
/*!
  * @file DFRobot_GP85XX.h
  * @brief 
  * @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  * @license     The MIT License (MIT)
  * @author      [Baikunlin](kunlin.bai@dfrobot.com)
  * @version  V1.0
  * @date  2023-02-10
  * @url https://github.com/DFRobot/DFRobot_GP85XX
  */
 
#include "DFRobot_GP85XX.h"

#define GP85XX_STORE_TIMING_HEAD            0x02  ///< Store function timing start head
#define GP85XX_STORE_TIMING_ADDR            0x10  ///< The first address for entering store timing
#define GP85XX_STORE_TIMING_CMD1            0x03  ///< The command 1 to enter store timing
#define GP85XX_STORE_TIMING_CMD2            0x00  ///< The command 2 to enter store timing
#define GP85XX_STORE_TIMING_DELAY           10    ///< Store procedure interval delay time: 10ms, more than 7ms
#define I2C_CYCLE_TOTAL                     5     ///< Total I2C communication cycle
#define I2C_CYCLE_BEFORE                    1     ///< The first half cycle 2 of the total I2C communication cycle
#define I2C_CYCLE_AFTER                     2     ///< The second half cycle 3 of the total I2C communication cycle
#define PWM_PASSAGG0                        0
#define PWM_PASSAGG1                        1

DFRobot_GP85XX::DFRobot_GP85XX()
{
}


/**************************************************************************
                       I2C转2路0-2.5/VCC 模拟电压模块(GP8503)
 **************************************************************************/
DFRobot_GP8503::DFRobot_GP8503(TwoWire *pWire,uint8_t addr)
{
  _pWire = pWire;
  _addr = addr;
}


void DFRobot_GP8503::begin(void)
{
  _pWire->begin();
}


void DFRobot_GP8503::setDACOutRange(eOutPutRange_t range,float vcc)
{
  if(range == eOutPutRange_t::eOutputRange2_5V)
  {
    _voltage = 2500;
  }else{
    _voltage = vcc*1000;
  }
}


void DFRobot_GP8503::setDACOutVoltage(uint16_t data, uint8_t channel,uint16_t value)
{
  if(data > _voltage){
    data = _voltage;
  }
  data = (uint16_t)round(((float)data / _voltage) * value);
  Serial.println(data);
  if(value == 4095 ){
    data = data << 4;
    sendData(data, channel);
  }else{
    sendData(data, 0);
  }
}


void DFRobot_GP8503::sendData(uint16_t data, uint8_t channel)
{
  if(channel == 0){
    _pWire->beginTransmission(_addr);
    _pWire->write(GP85XX_CONFIG_CURRENT_REG);
    _pWire->write(data & 0xff);
    _pWire->write((data >>8) & 0xff);
    _pWire->endTransmission();
  }else if(channel == 1){
	_pWire->beginTransmission(_addr);
    _pWire->write(GP85XX_CONFIG_CURRENT_REG<<1);
    _pWire->write(data & 0xff);
    _pWire->write((data >>8) & 0xff);
    _pWire->endTransmission();
  }else{
    _pWire->beginTransmission(_addr);
    _pWire->write(GP85XX_CONFIG_CURRENT_REG);
    _pWire->write(data & 0xff);
    _pWire->write((data >>8) & 0xff);
    _pWire->write(data & 0xff);
    _pWire->write((data >>8) & 0xff);
    _pWire->endTransmission();
  }
}


void DFRobot_GP8503::store(void)
{
  #if defined(ESP32)
    _pWire->~TwoWire();
  #elif !defined(ESP8266)
    _pWire->end();
  #endif
  
  pinMode(_scl, OUTPUT);
  pinMode(_sda, OUTPUT);
  digitalWrite(_scl, HIGH);
  digitalWrite(_sda, HIGH);
  startSignal();
  sendByte(GP85XX_STORE_TIMING_HEAD, 0, 3, false);
  stopSignal();
  startSignal();
  sendByte(GP85XX_STORE_TIMING_ADDR,0);
  sendByte(GP85XX_STORE_TIMING_CMD1,0);
  stopSignal();
  
  startSignal();
  sendByte(_addr<<1, 1);
  sendByte(GP85XX_STORE_TIMING_CMD2, 1);
  sendByte(GP85XX_STORE_TIMING_CMD2, 1);
  sendByte(GP85XX_STORE_TIMING_CMD2, 1);
  sendByte(GP85XX_STORE_TIMING_CMD2, 1);
  sendByte(GP85XX_STORE_TIMING_CMD2, 1);
  sendByte(GP85XX_STORE_TIMING_CMD2, 1);
  sendByte(GP85XX_STORE_TIMING_CMD2, 1);
  sendByte(GP85XX_STORE_TIMING_CMD2, 1);
  stopSignal();

  delay(GP85XX_STORE_TIMING_DELAY);

  startSignal();
  sendByte(GP85XX_STORE_TIMING_HEAD, 0, 3, false);
  stopSignal();
  startSignal();
  sendByte(GP85XX_STORE_TIMING_ADDR);
  sendByte(GP85XX_STORE_TIMING_CMD2);
  stopSignal();
  _pWire->begin();
}


void DFRobot_GP8503::startSignal(void)
{
  digitalWrite(_scl,HIGH);
  digitalWrite(_sda,HIGH);
  delayMicroseconds(I2C_CYCLE_BEFORE);
  digitalWrite(_sda,LOW);
  delayMicroseconds(I2C_CYCLE_AFTER);
  digitalWrite(_scl,LOW);
  delayMicroseconds(I2C_CYCLE_TOTAL);
}


void DFRobot_GP8503::stopSignal(void)
{
  digitalWrite(_sda,LOW);
  delayMicroseconds(I2C_CYCLE_BEFORE);
  digitalWrite(_scl,HIGH);
  delayMicroseconds(I2C_CYCLE_TOTAL);
  digitalWrite(_sda,HIGH);
  delayMicroseconds(I2C_CYCLE_TOTAL);
}


uint8_t DFRobot_GP8503::sendByte(uint8_t data, uint8_t ack, uint8_t bits, bool flag)
{
  for(int i=bits-1; i>=0;i--){  
    if(data & (1<<i)){
      digitalWrite(_sda,HIGH);
    }else{
      digitalWrite(_sda,LOW);
    }
    delayMicroseconds(I2C_CYCLE_BEFORE); 
    digitalWrite(_scl,HIGH);
    delayMicroseconds(I2C_CYCLE_TOTAL);
    digitalWrite(_scl,LOW);
    delayMicroseconds(I2C_CYCLE_AFTER); 
  }
  if(flag) return recvAck(ack);
  else {
    digitalWrite(_sda,LOW);
    digitalWrite(_scl,HIGH);
    return 0;
  }
}


uint8_t DFRobot_GP8503::recvAck(uint8_t ack)
{
  uint8_t ack_=0;
  uint16_t errorTime = 0;
  pinMode(_sda,INPUT_PULLUP);
  digitalWrite(_sda,HIGH);
  delayMicroseconds(I2C_CYCLE_BEFORE);
  digitalWrite(_scl,HIGH);
  delayMicroseconds(I2C_CYCLE_AFTER);
  while(digitalRead(_sda) != ack){
      delayMicroseconds(1);
      errorTime++;
      if(errorTime > 250) break;
  }
  ack_=digitalRead(_sda);
  delayMicroseconds(I2C_CYCLE_BEFORE);
  digitalWrite(_scl,LOW);
  delayMicroseconds(I2C_CYCLE_AFTER);
  pinMode(_sda,OUTPUT);
  return ack_;
}


/**************************************************************************
                       I2C转0-2.5/VCC (15bit) 模拟电压模块(GP8512)
 **************************************************************************/

DFRobot_GP8512::DFRobot_GP8512()
{
}
 
void DFRobot_GP8512::begin(void)
{
  _pgp8503.begin();
}


void DFRobot_GP8512::setDACOutRange(eOutPutRange_t range,float vcc)
{
  if(range == eOutputRange2_5V){
    _pgp8503.setDACOutRange(_pgp8503.eOutputRange2_5V,vcc);
  }else{
    _pgp8503.setDACOutRange(_pgp8503.eOutputRangeVCC,vcc);
  }
  
}


void DFRobot_GP8512::setDACOutVoltage(uint16_t data)
{
  _pgp8503.setDACOutVoltage(data,0,32767);
}


void DFRobot_GP8512::store(void)
{
  _pgp8503.store();
}


void DFRobot_GP8512::sendData(uint16_t data)
{
  _pgp8503.sendData(data,0);
}

/**************************************************************************
                       PWM转0-5V/VCC 模拟电压模块(GP8101)
 **************************************************************************/
 
void DFRobot_GP8101::begin(uint8_t pin)
{
  _pin = pin;
  #if defined(ESP32)  
    ledcSetup(PWM_PASSAGG0, 5000, 13);  ////分辨率为12位，即占空比可选0~8191
	ledcAttachPin(_pin, PWM_PASSAGG0);
  #elif defined(ESP8266) 
    analogWriteRange(8191);
  #else 
    pinMode(_pin,OUTPUT);
  #endif
}

void DFRobot_GP8101::setDACOutRange(eOutPutRange_t range)
{
   if(range == eOutPutRange_t::eOutputRange5V)
  {
    _voltage = 5000;
  }else{
    _voltage = 10000;
  }
}


void DFRobot_GP8101::setDACOutVoltage(uint16_t data)
{
  if(data > _voltage){
    data = _voltage;
  }
  #if defined(ESP32) || defined(ESP8266)
    data = (uint16_t)round(((float)data/_voltage)* 8191);
  #else 
    data = (uint16_t)round(((float)data/_voltage)* 255);
  #endif
  sendData(data); 
}


void DFRobot_GP8101::sendData(uint16_t data)
{
  #if defined(ESP32)  
    ledcWrite(PWM_PASSAGG0, data);
  #else 
    analogWrite(_pin, data);
  #endif
}
 
/**************************************************************************
                       PWM转2路0-2.5V/VCC 模拟电压模块(GP8501)
 **************************************************************************/


void DFRobot_GP8501::begin(uint8_t pin)
{
  _pin0 = pin;
   #if defined(ESP32)  
    ledcSetup(PWM_PASSAGG0, 5000, 13);  ////分辨率为12位，即占空比可选0~8191
	ledcAttachPin(_pin0, PWM_PASSAGG0);
  #elif defined(ESP8266) 
    analogWriteRange(8191);
  #else 
    pinMode(_pin0,OUTPUT);
  #endif
}


void DFRobot_GP8501::begin(uint8_t pin0,uint8_t pin1)
{
  _pin0 = pin0;
  _pin1 = pin1;
  #if defined(ESP32)  
    ledcSetup(PWM_PASSAGG0, 5000, 13);  ////分辨率为13位，即占空比可选0~8191
	ledcSetup(PWM_PASSAGG1, 5000, 13);
	ledcAttachPin(_pin0, PWM_PASSAGG0);
	ledcAttachPin(_pin1, PWM_PASSAGG1);
  #elif defined(ESP8266) 
    analogWriteRange(8191);
  #else 
    pinMode(_pin0,OUTPUT);
    pinMode(_pin1,OUTPUT);
  #endif
}


void DFRobot_GP8501::setDACOutRange(eOutPutRange_t range, float vcc)
{
  if(range == eOutPutRange_t::eOutputRange2_5V)
  {
    _voltage = 2500;
  }else{
    _voltage = vcc*1000;
  }
}


void DFRobot_GP8501::setDACOutVoltage(uint16_t data , uint8_t channel)
{ 
  if(data > _voltage){
    data = _voltage;
  }
  #if defined(ESP32) || defined(ESP8266)
    data = (uint16_t)round(((float)data/_voltage)* 8191);
  #else 
    data = (uint16_t)round(((float)data/_voltage)* 255);
  #endif
  sendData(data, channel);
}


void DFRobot_GP8501::sendData(uint16_t data, uint8_t channel)
{
  #if defined(ESP32)  
    if(channel == 0){
      ledcWrite(PWM_PASSAGG0, data);
    }else if(channel == 1){
      ledcWrite(PWM_PASSAGG1, data);
    }else{
      ledcWrite(PWM_PASSAGG0, data);
	  ledcWrite(PWM_PASSAGG1, data);
    }
  #else 
    if(channel == 0){
      analogWrite(_pin0, data);
    }else if(channel == 1){
      analogWrite(_pin1, data);
    }else{
      analogWrite(_pin0, data);
	  analogWrite(_pin1, data);
    }
  #endif
}